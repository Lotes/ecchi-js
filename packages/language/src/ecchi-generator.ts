import { URI, assertUnreachable } from "langium";
import { EcchiServices, createEcchiServices } from "./ecchi-module.js";
import {
  ConceptDefinition,
  Model,
  SubjectDefinition,
  TypeReference,
} from "./generated/ast.js";
import { EmptyFileSystem } from "langium";
import { readFile } from "fs/promises";
import {
  ConceptMap,
  EcchiGeneratorModel,
  SubjectData,
  buildGeneratorModel,
} from "./ecchi-generator-model.js";
import { OpcodeElement } from "./ecchi-generator-conditions.js";

export async function generate(fileName: string) {
  const services = createEcchiServices(EmptyFileSystem);
  const input = await readFile(fileName, "utf-8");
  const generator = services.Ecchi.generator.EcchiGenerator;
  const model = await generator.parse(input);
  const pkg = await generator.build(model);
  return await generator.generate(pkg);
}

export class EcchiGenerator {
  private nextDocumentId: number = 0;
  constructor(private readonly services: EcchiServices) {}
  async parse(ecchiFileContent: string) {
    const metaData = this.services.LanguageMetaData;
    const documentBuilder = this.services.shared.workspace.DocumentBuilder;
    const uri = URI.parse(
      `file:///${this.nextDocumentId++}${metaData.fileExtensions[0] ?? ""}`
    );
    const document =
      this.services.shared.workspace.LangiumDocumentFactory.fromString<Model>(
        ecchiFileContent,
        uri
      );
    this.services.shared.workspace.LangiumDocuments.addDocument(document);
    await documentBuilder.build([document], { validation: true });
    try {
      return document.parseResult.value;
    } finally {
      await documentBuilder.update([], [uri]);
    }
  }
  async build(model: Model) {
    return buildGeneratorModel(model);
  }
  toJSType(type: TypeReference): string {
    switch (type.$type) {
      case "BooleanType":
        return "boolean";
      case "NumberType":
        return "number";
      case "NullType":
        return "null";
      case "StringType":
        return "string";
      case "ArrayType":
        return `${this.toJSType(type.type)}[]`;
      case "ConceptReference":
        return type.type.ref!.name;
      default:
        assertUnreachable(type);
    }
  }
  async generate({ concepts, user, subjects }: EcchiGeneratorModel) {
    return `${this.generateImports()}

${this.generateTypes(concepts)}

${this.generateReflection(concepts)}

${this.generateUser(user)}

${this.generateSubjectActions(subjects)}
`;
  }
  generateConditions(pkg: EcchiGeneratorModel): string {
    return [...pkg.roles.entries()]
      .map(([role, data]) => {
        return `${this.generateImports()}

${this.generateTypes(pkg.concepts)}
${this.generateReflection(pkg.concepts)}
${this.generateUser(pkg.user)}

export function ${role.name}(user: $UserType) {
  const commonExpressions = [
    ${data.expressions.map(e => this.generateExpression(e)).join(",\n    ")}
  ] as const;
  return {
    ${role.members
      .map((member) => {
        const subjectName = member.subject.ref!.name;
        const subjectRules = data.rules.get(member.subject.ref!)!;
        const subjectType = member.subject.ref!.type.ref!.name;
        return `${subjectName}: (subject: ${subjectType}): [boolean, 'allow'|'forbid', string[]][] => {
      const subjectExpressions = [
        ${subjectRules.expressions.map(e => this.generateExpression(e)).join(",\n        ")}
      ] as const;
  
      return [
        ${[...subjectRules.rules]
          .map((action) => {
            return `[${action.condition <  0? `subjectExpressions[${-action.condition}]` :  `commonExpressions[${action.condition}]`}(), '${
              action.mode
            }', [${action.actions.map((e) => `'${e.name}'`).join(", ")}]]`;
          })
          .join(",\n        ")}
      ];
    }`;
      })
      .join(",\n    ")}
  }
}`;
      })
      .join(",\n    ");
  }
  generateExpression(expression: OpcodeElement): string {
    let code = "";
    const op = expression.code;
    function get(operandIndex: number) {
      return operandIndex > 0 
        ? `commonExpressions[${operandIndex}]()`
        : `subjectExpressions[${-operandIndex}]()`;
    }
    switch (op.op) {
      case "null":
        code = `null`;
        break;
      case "boolean":
        code = `${op.value}`;
        break;
      case "built-in":
        code = `${op.object}`;
        break;
      case "binary":
        code = `${get(op.leftOperandIndex)} ${op.operator} ${get(op.rightOperandIndex)}`;
        break;
      case "string":
        code = `"${op.value.replaceAll('"', '\\"')}"`;
        break;
      case "number":
        code = `${op.value}`;
        break;
      case "unary":
        code = `${op.operator}${get(op.operandIndex)}`;
        break;
      case "get-property":
        code = `${get(op.receiverOperandIndex)}.${op.property}`;
        break;
      case "is":
        code = `$Reflection.isSubTypeOf(${get(op.operandIndex)}.$type, '${op.type}')`;
        break;
      case "array-get":
        code = `${get(op.receiverOperandIndex)}[${get(op.indexOperandIndex)}]`;
        break;
      default:
        assertUnreachable(op);
    }
    const type = this.toJSType(expression.type);
    return `(): ${type} => ${code}`;
  }
  generateUser(user: ConceptDefinition | undefined) {
    return `export type $UserType = ${user?.name};`;
  }
  generateSubjectActions(map: Map<SubjectDefinition, SubjectData>) {
    return `export const $SubjectActions = {
  ${[...map.keys()]
    .map((subject) => {
      const { parents, hierarchy: nestedSets, instances } = map.get(subject)!;
      const type = [...parents.keys()].map((a) => `'${a}'`).join("|");
      return `${subject.name}: ["${
        subject.type?.ref?.name
      }", new SubjectActions<${type}>({
    ${[...instances.entries()]
      .map(([action, tree], index) => {
        const [left, right] = nestedSets.get(tree.content)!;
        return `${action}: [[${left}, ${right}],  ${index}]`;
      })
      .join(",\n    ")}
  })]`;
    })
    .join(",\n  ")}
} satisfies SubjectActionsBase<$Types>;    
`;
  }
  private generateReflection(concepts: ConceptMap) {
    return `export const $Reflection = new Reflection<$Types>({
  ${[...concepts.entries()]
    .map(([concept, { nestedSet: set }]) => {
      return `${concept.name}: [${set[0]},  ${set[1]}],`;
    })
    .join("\n  ")}
});`;
  }

  public generateTypes(concepts: ConceptMap) {
    function visit(concept: ConceptDefinition) {
      const subTypes: string = concepts
        .get(concept)!
        .children.map(visit)
        .join("|");
      return `"${concept.name}"${subTypes.length > 0 ? `|${subTypes}` : ""}`;
    }
    return (
      [...concepts.keys()]
        .map((type) => {
          return `interface ${type.name}${
            type.superConcept ? ` extends ${type.superConcept.ref?.name}` : ""
          } {
  $type: ${visit(type)};
  ${type.members
    .map(
      (member) => `${member.name}: ${this.generateTypeReference(member.type)};`
    )
    .join("\n  ")}
}`;
        })
        .join("\n") +
      `

export type $Types = {
  ${[...concepts.keys()]
    .map((type) => {
      return `${type.name}: ${type.name},`;
    })
    .join("\n  ")}
}`
    );
  }
  private generateTypeReference(type: TypeReference): string {
    switch (type.$type) {
      case "BooleanType":
        return "boolean";
      case "NumberType":
        return "number";
      case "NullType":
        return "null";
      case "StringType":
        return "string";
      case "ArrayType":
        return `${this.generateTypeReference(type.type)}[]`;
      case "ConceptReference":
        return type.type.ref!.name;
      default:
        assertUnreachable(type);
    }
  }
  private generateImports() {
    return `import { Reflection, RoleRules, SubjectActions, SubjectActionsBase, not } from "@ecchi-js/core";`;
  }
}
