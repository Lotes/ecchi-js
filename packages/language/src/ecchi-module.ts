import type {
  DefaultSharedModuleContext,
  LangiumServices,
  LangiumSharedServices,
  Module,
  PartialLangiumServices,
} from "langium";
import {
  createDefaultModule,
  createDefaultSharedModule,
  inject,
} from "langium";
import {
  EcchiGeneratedModule,
  EcchiGeneratedSharedModule,
} from "./generated/module.js";
import { EcchiValidator, registerValidationChecks } from "./ecchi-validator.js";
import { EcchiScopeProvider } from "./ecchi-scope-provider.js";
import { EcchiGenerator } from "./ecchi-generator.js";

/**
 * Declaration of custom services - add your own service classes here.
 */
export type EcchiAddedServices = {
  validation: {
    EcchiValidator: EcchiValidator;
  };
  generator: {
    EcchiGenerator: EcchiGenerator;
  };
};

/**
 * Union of Langium default services and your custom services - use this as constructor parameter
 * of custom service classes.
 */
export type EcchiServices = LangiumServices & EcchiAddedServices;

/**
 * Dependency injection module that overrides Langium default services and contributes the
 * declared custom services. The Langium defaults can be partially specified to override only
 * selected services, while the custom services must be fully specified.
 */
export const EcchiModule: Module<
  EcchiServices,
  PartialLangiumServices & EcchiAddedServices
> = {
  references: {
    ScopeProvider: (services) => new EcchiScopeProvider(services),
  },
  validation: {
    EcchiValidator: () => new EcchiValidator(),
  },
  generator: {
    EcchiGenerator: (services) => new EcchiGenerator(services),
  },
};

/**
 * Create the full set of services required by Langium.
 *
 * First inject the shared services by merging two modules:
 *  - Langium default shared services
 *  - Services generated by langium-cli
 *
 * Then inject the language-specific services by merging three modules:
 *  - Langium default language-specific services
 *  - Services generated by langium-cli
 *  - Services specified in this file
 *
 * @param context Optional module context with the LSP connection
 * @returns An object wrapping the shared services and the language-specific services
 */
export function createEcchiServices(context: DefaultSharedModuleContext): {
  shared: LangiumSharedServices;
  Ecchi: EcchiServices;
} {
  const shared = inject(
    createDefaultSharedModule(context),
    EcchiGeneratedSharedModule
  );
  const Ecchi = inject(
    createDefaultModule({ shared }),
    EcchiGeneratedModule,
    EcchiModule
  );
  shared.ServiceRegistry.register(Ecchi);
  registerValidationChecks(Ecchi);
  return { shared, Ecchi };
}
