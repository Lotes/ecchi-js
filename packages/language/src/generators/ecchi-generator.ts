import { URI, assertUnreachable } from "langium";
import { EcchiServices } from "../ecchi-module.js";
import { InterfaceDefinition, Model, SubjectDefinition, TypeReference, UserDefinition, UserMemberDefinition, isInterfaceDefinition, isSubjectDefinition, isUserDefinition } from "../generated/ast.js";
import { NestedSetElement } from "@ecchi-js/core";

interface TypeTree {
  $type: string;
  children: TypeTree[];
}

export class EcchiGenerator {
  private nextDocumentId: number = 0;
  constructor(private readonly services: EcchiServices) {}
  async generate(ecchiFileContent: string) {
    const metaData = this.services.LanguageMetaData;
    const documentBuilder = this.services.shared.workspace.DocumentBuilder;
    const uri = URI.parse(`file:///${this.nextDocumentId++}${metaData.fileExtensions[0] ?? ''}`);
    const document = this.services.shared.workspace.LangiumDocumentFactory.fromString<Model>(ecchiFileContent, uri);
    this.services.shared.workspace.LangiumDocuments.addDocument(document);
    await documentBuilder.build([document], { validation: true });
    try {
      return this.generateFromModel(document.parseResult.value);
    } finally {
      await documentBuilder.update([], [uri]);
    }
  }
  private generateFromModel(model: Model) {
    const { interfaces, nestedSets, trees } = this.getInterfaces(model);
    const dts = `${this.generateImports()}

${this.generateTypes(interfaces, trees)}

${this.generateReflection(interfaces, nestedSets)}

${this.generateUser(model.elements.filter(isUserDefinition))}

${this.generateSubjectActions(model.elements.filter(isUserDefinition).flatMap(u => u.members))}
`;
    return dts;
  }
  generateUser(users: UserDefinition[]) {
    return `export type $UserType = ${users[0]?.type?.ref?.name};`;
  }
  private getSubjectActions(member: SubjectDefinition) {
    const parents: Record<string, string|undefined> = Object.fromEntries(member.members.map(m => [m.name, m.superAction?.ref?.name] as const));
    const children: Record<string, string[]> = Object.fromEntries(Object.entries(parents).map(([name]) => [name, [] as string[]] as const));
    for (const [child, parent] of Object.entries(parents).filter(([_, parent]) => parent !== undefined)) {
      children[parent!].push(child);
    }
    return {
      parents,
      children
    };
  }
  generateSubjectActions(members: UserMemberDefinition[]) {
    const subjects = members.filter(isSubjectDefinition);
    return `export const $SubjectActions = {
  ${subjects.map(subject => {
    const { children, parents } = this.getSubjectActions(subject);
    const type = Object.keys(parents).map(a => `'${a}'`).join('|');
    const nestedSets = new Map<string, NestedSetElement>();
    let counter = 0;
    function visit(node: string) {
      const left = counter++;
      for (const child of children[node]) {
        visit(child);
      }
      const right = counter++;
      nestedSets.set(node, [
        left,
        right
      ]);
    }
    for (const node of Object.entries(parents).filter(([_, parent]) => parent === undefined).map(([action]) => action)){
      visit(node);
    }
    return `${subject.name}: ["${subject.type.ref?.name}", new SubjectActions<${type}>({
    ${Object.keys(parents).map((action, index) => {
        const [left, right] = nestedSets.get(action)!;
        return `${action}: [[${left}, ${right}],  ${index}]`;
      }).join(',\n    ')}
  })]`; 
  }).join('\n  ')}
} satisfies SubjectActionsBase<$Types>;    
`;
  }
  private generateReflection(interfaces: InterfaceDefinition[], nestedSets: Map<string, NestedSetElement>) {
    return `export const $Reflection = new Reflection<$Types>({
  ${interfaces.map(iface => {
  const set = nestedSets.get(iface.name)!;
  return `${iface.name}: [${set[0]},  ${set[1]}],`
}).join('\n  ')}
});`;
  }
  private getInterfaces(model: Model) {
    const map = new Map<string, TypeTree>();
    const nestedSets = new Map<string, NestedSetElement>();
    const interfaces = model.elements.filter(isInterfaceDefinition);
    const roots: TypeTree[] = [];
    for (const iface of interfaces) {
      map.set(iface.name, {
        $type: iface.name,
        children: []
      });
    }
    for (const iface of interfaces) {
      const node = map.get(iface.name)!;
      if (iface.superInterface) {
        const superNode = map.get(iface.superInterface.ref!.name)!;
        superNode.children.push(node);
      } else {
        roots.push(node);
      }
    }
    let counter = 0;
    function visit(node: TypeTree) {
      const left = counter++;
      for (const child of node.children) {
        visit(child);
      }
      const right = counter++;
      nestedSets.set(node.$type, [
        left,
        right
      ]);
    }
    for (const node of roots) {
      visit(node);
    }
    return { interfaces, nestedSets, trees: map };
  }

  public generateTypes(interfaces: InterfaceDefinition[], trees: Map<string, TypeTree>) {
    function visit(name: string) {
      const subTypes: string = trees.get(name)!.children.map(c => visit(c.$type)).join('|');
      return `"${name}"${subTypes.length > 0 ? `|${subTypes}` : ''}`;
    }
    return interfaces.map(type => { 
      return `interface ${type.name}${type.superInterface ? ` extends ${type.superInterface.ref?.name}` : ''} {
  $type: ${visit(type.name)};
  ${type.members.map(member => `${member.name}: ${this.generateTypeReference(member.type)};`).join('\n  ')}
}`;
}).join('\n')+`

export type $Types = {
  ${interfaces.map(type => {
  return `${type.name}: ${type.name},`;
}).join('\n  ')}
}`;
  }
  private generateTypeReference(type: TypeReference): string {
    switch (type.$type) {
      case 'BooleanType': return 'boolean';
      case 'NumberType': return 'number';
      case 'StringType': return 'string';
      case 'TypeDefinitionReference': return type.type.ref!.name;
      case 'ArrayType': return `${this.generateTypeReference(type.type)}[]`;
      case "NullType": return 'null';
      case "ObjectType": return `{
  ${type.members.map(member => `${member.name}: ${this.generateTypeReference(member.type)};`).join('\n  ')
}`;
      default: assertUnreachable(type);
    }
  }
  private generateImports() {
    return `import { Reflection, RoleRules, SubjectActions, SubjectActionsBase, not } from "@ecchi-js/core";`;
  }
}