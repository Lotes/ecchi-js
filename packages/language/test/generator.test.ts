import { test, expect } from "vitest";
import { EcchiGenerator } from "../src/ecchi-generator.js";
import { createEcchiServices } from "../src/ecchi-module.js";
import { EmptyFileSystem } from "langium";
import { readFile } from "fs/promises";
import { join } from "path";

test("ecchi-generator", async () => {
  const services = createEcchiServices(EmptyFileSystem);
  const generator = new EcchiGenerator(services.Ecchi);
  const content = await readFile(join(__dirname, "..", "..", "..", "resources", "Blog.ecchi"), "utf-8");
  const result = await generator.generate(content);
  expect(result).toMatchSnapshot();
});