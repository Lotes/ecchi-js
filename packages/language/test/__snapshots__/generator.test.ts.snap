// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ecchi-generator > blog 1`] = `
"import { Reflection, Cache, LRUCache, cacheCommonExpressions, cacheSubjectExpressions, merge } from "@ecchi-js/core";

interface UserType {
  $type: "UserType";
  id: number;
  name: string;
  email: string;
  createdAt: string;
  updatedAt: string;
}
interface ArticleType {
  $type: "ArticleType";
  id: number;
  title: string;
  content: string;
  createdAt: string;
  updatedAt: string;
  author: UserType;
  published: boolean;
}
interface CommentType {
  $type: "CommentType";
  id: number;
  content: string;
  createdAt: string;
  updatedAt: string;
  author: UserType;
  article: ArticleType;
  published: boolean;
}
interface A {
  $type: "A"|"B";
  
}
interface B extends A {
  $type: "B";
  
}

export type $Types = {
  UserType: UserType,
  ArticleType: ArticleType,
  CommentType: CommentType,
  A: A,
  B: B,
}

export const $Reflection = new Reflection<$Types>({
  UserType: [0,  1],
  ArticleType: [2,  3],
  CommentType: [4,  5],
  A: [6,  9],
  B: [7,  8],
});

export type $UserType = UserType;


export type $Role = 'AdminUser'|'NormalUser';
export type $Subject = 'Article'|'Comment';
export type $Subjects = {
  Article: ArticleType;
  Comment: CommentType;
};
export type $Actions = {
  Article: 'create'|'read'|'edit'|'archive'|'publish'|'delete';
  Comment: 'read'|'answer'|'edit'|'archive'|'delete'|'publish';
};

export type CanOptions = {
  I: $UserType;
  actingAs?: $Role[];
  cache?: Cache;
} & ({
  when: 'Article';
  subject: ArticleType;
  doWhat: $Actions['Article'];
} |  {
  when: 'Comment';
  subject: CommentType;
  doWhat: $Actions['Comment'];
});

const DefaultCache = new LRUCache(128);

export function can({
  I: user,
  actingAs = [],
  when,
  subject,
  doWhat,
  cache = DefaultCache
}: CanOptions) {
  type Common = readonly [null, null, boolean, boolean, UserType, number];
  const commonExpressions = cacheCommonExpressions<Common>([
    (): null => null,
    (): null => null,
    (): boolean => true,
    (): boolean => false,
    (): UserType => user,
    (commons): number => commons[4].id
  ] as const, [user], cache);

  const subjectHandlers: { [K in $Subject]: (subject: $Subjects[K]) => boolean} = {
    Article(subject: ArticleType) {
      type Subject = readonly [null, ArticleType, UserType, number, boolean, boolean, boolean, boolean];
      const subjectExpressions = cacheSubjectExpressions<Common, Subject>(commonExpressions, [
        (): null => null,
        (): ArticleType => subject,
        (_common, subjects): UserType => subjects[1].author,
        (_common, subjects): number => subjects[2].id,
        (commons, subjects): boolean => commons[5] == subjects[3],
        (commons, subjects): boolean => commons[2] && subjects[4],
        (_common, subjects): boolean => !subjects[4],
        (commons, subjects): boolean => commons[2] && subjects[6]
      ] as const, [subject], cache);
      const actionBits = {
        create: {
          allow: [0, 1 << 0],
          forbid: [0, 1 << 1],
        },
        read: {
          allow: [0, 1 << 2],
          forbid: [0, 1 << 3],
        },
        edit: {
          allow: [0, 1 << 4],
          forbid: [0, 1 << 5],
        },
        archive: {
          allow: [0, 1 << 6],
          forbid: [0, 1 << 7],
        },
        publish: {
          allow: [1, 1 << 0],
          forbid: [1, 1 << 1],
        },
        delete: {
          allow: [1, 1 << 2],
          forbid: [1, 1 << 3],
        },
      };
      const roleHandlers: Record<$Role, (() => [boolean, number[]])[]> = {
        AdminUser: [
          () => [commonExpressions[2], [0x55, 0x05]]
        ],
        NormalUser: [
          () => [subjectExpressions[5], [0x54, 0x00]],
          () => [subjectExpressions[7], [0x04, 0x00]]
        ],
      };
      const { allow, forbid } = actionBits[doWhat];
      return merge(allow, forbid, actingAs, roleHandlers);
    },
    Comment(subject: CommentType) {
      type Subject = readonly [null];
      const subjectExpressions = cacheSubjectExpressions<Common, Subject>(commonExpressions, [
        (): null => null
      ] as const, [subject], cache);
      const actionBits = {
        read: {
          allow: [0, 1 << 0],
          forbid: [0, 1 << 1],
        },
        answer: {
          allow: [0, 1 << 2],
          forbid: [0, 1 << 3],
        },
        edit: {
          allow: [0, 1 << 4],
          forbid: [0, 1 << 5],
        },
        archive: {
          allow: [0, 1 << 6],
          forbid: [0, 1 << 7],
        },
        delete: {
          allow: [1, 1 << 0],
          forbid: [1, 1 << 1],
        },
        publish: {
          allow: [1, 1 << 2],
          forbid: [1, 1 << 3],
        },
      };
      const roleHandlers: Record<$Role, (() => [boolean, number[]])[]> = {
        AdminUser: [
          () => [commonExpressions[2], [0x55, 0x05]]
        ],
        NormalUser: [
          
        ],
      };
      const { allow, forbid } = actionBits[doWhat];
      return merge(allow, forbid, actingAs, roleHandlers);
    },
  };
  return subjectHandlers[when](subject as any);
}
"
`;
